compiler/util/misc.cpp

```diff
@@ -251,53 +251,139 @@ static FnSymbol* findNonTaskCaller(FnSymbol* fn) {
   return lastFn; // never reached
 }
 
-// Print instantiation information for err_fn.
-// Should be called at USR_STOP or just before the next
-// error changing err_fn is printed.
-static void printInstantiationNoteForLastError() {
-  if (err_fn_header_printed && err_fn && err_fn->instantiatedFrom) {
-
-    // Find the first call to the function within the instantiation point,
-    // so that we can have a better error message line number.
-    BlockStmt* instantiationPoint = err_fn->instantiationPoint();
-    Expr* bestPoint = instantiationPoint;
-
-    if (instantiationPoint != NULL) {
-      std::vector<CallExpr*> calls;
-      collectFnCalls(instantiationPoint, calls);
-      for_vector(CallExpr, call, calls) {
-        if (FnSymbol* fn = call->resolvedOrVirtualFunction()) {
-          if (fn == err_fn) {
-            bestPoint = call;
-            break;
-          }
+static void gatherFunctionsCalledTransitively(FnSymbol* fn,
+                                              std::set<FnSymbol*>& fns) {
+  if (fns.count(fn) == 0) {
+    fns.insert(fn);
+
+    // Gather calls from the function body & declaration
+    std::vector<CallExpr*> calls;
+    collectVirtualAndFnCalls(fn, calls);
+
+    // gather calls from the called functions (transitively)
+    for_vector(CallExpr, call, calls) {
+      FnSymbol* calledFn = call->resolvedOrVirtualFunction();
+      gatherFunctionsCalledTransitively(calledFn, fns);
+    }
+  }
+}
+
+static CallExpr* findACallSite(FnSymbol* fn, std::set<FnSymbol*> ignoreFns) {
+  fn = findNonTaskCaller(fn);
+
+  for_SymbolSymExprs(se, fn) {
+    CallExpr* call = toCallExpr(se->parentExpr);
+    if (se == call->baseExpr) {
+      FnSymbol* inFn = call->getFunction();
+      if (inFn && ignoreFns.count(inFn) == 0)
+        return call;
+    }
+  }
+
+  return NULL;
+}
+
+static bool isModuleInitFunction(FnSymbol* fn) {
+  ModuleSymbol* mod = fn->getModule();
+  if (mod && mod->initFn == fn)
+    return true;
+  if (fn->hasFlag(FLAG_MODULE_INIT))
+    return true;
+
+  return false;
+}
+
+
+static const char* fnKindAndName(FnSymbol* fn) {
+  if (fn == NULL)
+    return "";
+
+  ModuleSymbol* mod = fn->getModule();
+  if (mod && mod->initFn == fn)
+    return astr("module ", "'", mod->name, "'");
+
+  if (strcmp(fn->name, "init") == 0)
+    return astr("initializer");
+
+  if (fn->isIterator())
+    return astr("iterator ", "'", fn->name, "'");
+
```

Comment:
Should we also differentiate "method" ? Arguably it is in the same category as "initializer"